core/domain/classifier_services.py

```diff
@@ -127,6 +127,69 @@ def classify_string_classifier_rule(state, normalized_answer):
     return None
 
 
+
+
+def add_to_training_queue(exploration):
```

Comment:
The function name seems misleading -- there's no actual training queue, right? Also, looking at what it does, it calls is_retraining_necessary() for each state. This is expensive (each call involves a datastore GET operation), and probably not actually needed if this is meant to be called when an exploration is saved. 

Could this function become something like create_classifier_training_jobs(exploration_id, state_names) instead? You can leave exp_services._save_exploration() or similar to do the exploration-level processing since it presumably has the old/new versions of the exploration already, so hopefully it can compare those directly and just pass the list of states whose answer groups have changed. Then, somewhere in the "save exploration" flow, you might do the following (feel free to tweak the naming):

    state_names_with_changed_answer_groups = exploration.get_state_names_with_changed_answer_groups(old_states)
    state_names_with_unchanged_answer_groups = set(exploration.states.keys()) - set(state_names_with_changed_answer_groups)

    classifier_services.create_training_jobs(exploration_id, state_names_with_changed_answer_groups)
    classifier_services.update_classifier_mappings(exploration_id, state_names_with_unchanged_answer_groups)

That might be a better separation of concerns.